service/udp_linux.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2018 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package service
+
+import (
+	"context"
+	"net"
+	"syscall"
+	"unsafe"
+
+	onet "github.com/Jigsaw-Code/outline-ss-server/net"
+	"golang.org/x/sys/unix"
+)
+
+func ListenUDP(network string, laddr *net.UDPAddr) (onet.UDPPacketConn, error) {
+	lc := &net.ListenConfig{
+		Control: func(network, address string, c syscall.RawConn) error {
+			return c.Control(func(fd uintptr) {
+				// Always set IP_PKTINFO
+				if err := unix.SetsockoptInt(int(fd), unix.IPPROTO_IP, unix.IP_PKTINFO, 1); err != nil {
+					logger.Errorf("failed to set socket option IP_PKTINFO: %v", err)
+					return
+				}
+				logger.Debugf("successfully set IP_PKTINFO on %s", network)
+
+				if network == "udp6" {
+					if err := unix.SetsockoptInt(int(fd), unix.IPPROTO_IPV6, unix.IPV6_RECVPKTINFO, 1); err != nil {
+						logger.Errorf("failed to set socket option IPV6_RECVPKTINFO: %v", err)
+						return
+					}
+					logger.Debugf("successfully set IPV6_RECVPKTINFO on %s", network)
+				}
+			})
+		},
+	}
+
+	conn, err := lc.ListenPacket(context.Background(), network, laddr.String())
+	if err != nil {
+		return nil, err
+	} else {
+		return conn.(onet.UDPPacketConn), err
+	}
+}
+
+func getOobForCache(clientOob []byte) []byte {
```

Comment:
Can't you get the address with `PacketConn.LocalAddr()`?
https://pkg.go.dev/net#PacketConn

Reply:
You can't. Because our `clientConn` listens on `::`.

Reply:
I haven't looked into this in great detail, but I think the cleanest solution here would be to launch a separate `UDPService` for each network interface, each holding a `clientConn` bound to one server IP address.